File: avitowatcher/engine.py
"""Движок слежения: фоновый поток, который обходит задачи по расписанию.

Один поток на всё приложение — так запросы к Авито идут строго по очереди
с паузами между ними. Результаты передаются в интерфейс через сигналы Qt,
они безопасно пересекают границу потоков.
"""
from __future__ import annotations

import logging
import queue
import re
import threading
import time
import urllib.parse

from PySide6.QtCore import QObject, Signal

from . import notify, similarity
from .config import SettingsStore
from .db import Database
from .fetcher import BlockedError, Fetcher, RateLimiter, is_empty_result
from .models import (
    KIND_SEARCH,
    KIND_SIMILAR,
    STATUS_BLOCKED,
    STATUS_ERROR,
    STATUS_OK,
    STATUS_RUNNING,
    Listing,
    Task,
)
from .parser import ItemInfo, is_removed_item, parse_item, parse_search

log = logging.getLogger(__name__)

MIN_INTERVAL = 60  # чаще раза в минуту Авито дёргать не будем


def normalize_search_url(url: str) -> str:
    """Приводит ссылку поиска к виду «сначала свежие, первая страница»."""
    url = url.strip()
    if not url:
        return ""
    if not url.startswith("http"):
        url = "https://" + url.lstrip("/")
    parts = urllib.parse.urlsplit(url)
    params = urllib.parse.parse_qsl(parts.query, keep_blank_values=True)
    params = [(k, v) for k, v in params if k != "p"]  # всегда первая страница
    if not any(k == "s" for k, _ in params):
        params.append(("s", "104"))  # сортировка по дате
    return urllib.parse.urlunsplit(
        (parts.scheme, parts.netloc, parts.path, urllib.parse.urlencode(params), "")
    )


_ITEM_URL = re.compile(r"[-_](\d{6,})/?$")


def looks_like_item_url(url: str) -> bool:
    """Ссылка на конкретное объявление (в конце числовой номер)?"""
    path = urllib.parse.urlsplit(url.strip()).path
    return bool(_ITEM_URL.search(path)) and len([p for p in path.split("/") if p]) >= 2


def fetch_item_info(fetcher: Fetcher, url: str) -> ItemInfo:
    """Загружает объявление-образец и разбирает его (для задачи «похожие»)."""
    html = fetcher.get(url)
    if is_removed_item(html):
        raise RuntimeError(
            "Это объявление снято с публикации, взять его за образец нельзя. "
            "Выберите другое, похожее объявление."
        )
    info = parse_item(html, url)
    if not info.title:
        raise RuntimeError(
            "Не удалось прочитать объявление. Проверьте ссылку — нужна ссылка "
            "на страницу конкретного объявления."
        )
    return info


class WatcherEngine(QObject):
    """Планировщик проверок. Запускается и останавливается из интерфейса."""

    task_started = Signal(int)                  # task_id
    task_finished = Signal(int, str, str)       # task_id, статус, текст ошибки
    new_listings = Signal(int, list)            # task_id, список Listing
    notify_errors = Signal(list)                # проблемы с каналами уведомлений
    message = Signal(str)                       # строка для журнала в интерфейсе
    running_changed = Signal(bool)

    def __init__(self, db: Database, settings_store: SettingsStore) -> None:
        super().__init__()
        self._db = db
        self._settings = settings_store
        self._limiter = RateLimiter()
        self._fetcher = Fetcher(settings_store.current, self._limiter)
        self._thread: threading.Thread | None = None
        self._stop = threading.Event()
        self._wake = threading.Event()
        self._manual: queue.Queue[int] = queue.Queue()
        self._next_run: dict[int, float] = {}
        self._last_prune = 0.0

    # --- управление -------------------------------------------------------
    @property
    def is_running(self) -> bool:
        return self._thread is not None and self._thread.is_alive()

    def start(self) -> None:
        if self.is_running:
            return
        self._stop.clear()
        self._fetcher.update_settings(self._settings.current)
        self._thread = threading.Thread(target=self._loop, name="watcher", daemon=True)
        self._thread.start()
        self.running_changed.emit(True)
        self.message.emit("Слежение запущено")

    def stop(self, timeout: float = 5.0) -> None:
        if not self.is_running:
            return
        self._stop.set()
        self._wake.set()
        if self._thread is not None:
            self._thread.join(timeout=timeout)
        self._thread = None
        self.running_changed.emit(False)
        self.message.emit("Слежение остановлено")

    def shutdown(self) -> None:
        self.stop(timeout=3.0)
        self._fetcher.close()

    def settings_changed(self) -> None:
        self._fetcher.update_settings(self._settings.current)
        self._wake.set()

    def check_now(self, task_id: int) -> None:
        """Внеочередная проверка задачи."""
        self._manual.put(task_id)
        self._wake.set()

    def schedule_soon(self, task_id: int, delay: float = 2.0) -> None:
        self._next_run[task_id] = time.time() + delay
        self._wake.set()

    def forget_task(self, task_id: int) -> None:
        self._next_run.pop(task_id, None)

    @property
    def fetcher(self) -> Fetcher:
        return self._fetcher

    # --- основной цикл ----------------------------------------------------
    def _loop(self) -> None:
        while not self._stop.is_set():
            try:
                self._tick()
            except Exception:
                log.exception("Сбой в цикле слежения")
            self._wake.wait(timeout=2.0)
            self._wake.clear()

    def _tick(self) -> None:
        now = time.time()

        # Внеочередные проверки — вперёд расписания.
        while not self._manual.empty():
            try:
                task_id = self._manual.get_nowait()
            except queue.Empty:
                break
            task = self._db.get_task(task_id)
            if task is not None:
                self._run_task(task, manual=True)
            if self._stop.is_set():
                return

        for task in self._db.list_tasks():
            if self._stop.is_set():
                return
            if not task.enabled:
                continue
            interval = max(MIN_INTERVAL, task.interval)
            due = self._next_run.get(task.id, task.last_check + interval)
            if now >= due:
                self._run_task(task)

        if now - self._last_prune > 3600:
            self._last_prune = now
            settings = self._settings.current
            try:
                self._db.prune(settings.keep_history_days, settings.max_feed_items * 4)
            except Exception:
                log.exception("Не удалось почистить базу")

    # --- проверка одной задачи -------------------------------------------
    def _run_task(self, task: Task, manual: bool = False) -> None:
        settings = self._settings.current
        self._db.set_task_status(task.id, STATUS_RUNNING)
        self.task_started.emit(task.id)

        status, error = STATUS_OK, ""
        try:
            self._check_task(task, settings)
        except BlockedError as exc:
            status, error = STATUS_BLOCKED, str(exc)
            log.warning("Задача %s: блокировка — %s", task.name, exc)
        except Exception as exc:
            status, error = STATUS_ERROR, str(exc)
            log.exception("Задача %s: ошибка", task.name)

        task.last_check = time.time()
        task.status = status
        task.last_error = error
        self._db.update_task(task)
        self._next_run[task.id] = task.last_check + max(MIN_INTERVAL, task.interval)
        self.task_finished.emit(task.id, status, error)
        if error:
            self.message.emit(f"«{task.name}»: {error}")

    def _check_task(self, task: Task, settings) -> None:
        if task.kind == KIND_SIMILAR:
            self._ensure_similar_ready(task)
        search_url = normalize_search_url(task.search_url)
        if not search_url:
            raise RuntimeError("У задачи не задана ссылка для поиска")

        html = self._fetcher.get(
            search_url, stop_event=self._stop, require='data-marker="item"'
        )
        log.info(
            "Задача %s: страница получена (%s)", task.name, self._fetcher.last_transport
        )
        found = parse_search(html)
        if not found:
            if is_empty_result(html):
                # По запросу пока просто ничего нет — ждём дальше.
                task.params["baseline_done"] = True
                return
            raise RuntimeError(
                "Авито вернул страницу без объявлений. Проверьте ссылку "
                "или попробуйте позже."
            )

        candidates = [l for l in found if self._passes_filters(task, l)]
        first_run = not task.params.get("baseline_done")
        new_ids = self._db.filter_new(task.id, [l.item_id for l in candidates])

        # Первый проход — запоминаем текущую выдачу молча, иначе пользователь
        # получит полсотни уведомлений о давно висящих объявлениях.
        if first_run and not task.params.get("notify_existing"):
            self._db.mark_seen(task.id, [l.item_id for l in found])
            task.params["baseline_done"] = True
            self.message.emit(
                f"«{task.name}»: запомнил {len(candidates)} текущих объявлений, "
                f"жду новые"
            )
            return

        task.params["baseline_done"] = True
        fresh = [l for l in candidates if l.item_id in new_ids]
        self._db.mark_seen(task.id, [l.item_id for l in found])
        if not fresh:
            return

        for listing in fresh:
            listing.task_id = task.id
            listing.task_name = task.name
        task.found_total += len(fresh)

        self._db.add_listings(fresh)
        self.new_listings.emit(task.id, fresh)
        self.message.emit(f"«{task.name}»: новых объявлений — {len(fresh)}")
        log.info("Задача %s: новых объявлений %d", task.name, len(fresh))

        errors = notify.dispatch(settings, fresh, task.name)
        if errors:
            self.notify_errors.emit(errors)

    # Сколько слов заголовка пробуем оставить в запросе, от точного к широкому.
    QUERY_WIDTHS = (4, 3, 2)
    ENOUGH_NEIGHBOURS = 3

    def _ensure_similar_ready(self, task: Task) -> None:
        """Для задачи «похожие» один раз вычисляет поисковую ссылку."""
        if task.params.get("derived_url"):
            return
        info = fetch_item_info(self._fetcher, task.url)
        task.params["source_title"] = info.title
        task.params["source_price"] = info.price
        task.params["source_id"] = info.item_id
        task.params["derived_url"] = self._pick_search_url(task, info)
        self._db.update_task(task)
        self.message.emit(
            f"«{task.name}»: образец — {info.title} "
            f"(запрос «{task.params.get('derived_query', '')}»)"
        )

    def _pick_search_url(self, task: Task, info: ItemInfo) -> str:
        """Подбирает ширину запроса, чтобы в выдаче были не только образец.

        Полный заголовок объявления в качестве запроса часто находит ровно
        одно объявление — то самое, с которого мы начали. Поэтому один раз,
        при заведении задачи, пробуем запрос покороче, пока в выдаче не
        появятся соседи.
        """
        fallback_url, fallback_query = "", ""
        for width in self.QUERY_WIDTHS:
            query = similarity.build_query(info.title, width)
            if not query:
                continue
            url = similarity.build_search_url(info, task.price_tolerance, query)
            if not fallback_url:
                fallback_url, fallback_query = url, query
            try:
                html = self._fetcher.get(
                    url, stop_event=self._stop, require='data-marker="item"'
                )
            except Exception as exc:
                log.info("Проба запроса «%s» не удалась: %s", query, exc)
                continue
            neighbours = [
                item for item in parse_search(html)
                if item.item_id != info.item_id
            ]
            log.info("Запрос «%s»: соседей %d", query, len(neighbours))
            if len(neighbours) >= self.ENOUGH_NEIGHBOURS:
                task.params["derived_query"] = query
                return url
            fallback_url, fallback_query = url, query
        task.params["derived_query"] = fallback_query
        return fallback_url

    def _passes_filters(self, task: Task, listing: Listing) -> bool:
        title = (listing.title or "").lower()
        for word in task.keywords_exclude:
            if word in title:
                return False

        if task.kind == KIND_SEARCH:
            if task.price_min is not None and listing.price is not None:
                if listing.price < task.price_min:
                    return False
            if task.price_max is not None and listing.price is not None:
                if listing.price > task.price_max:
                    return False
            return True

        # «Похожие»: само объявление-образец в ленту не пускаем.
        if listing.item_id and listing.item_id == str(task.params.get("source_id", "")):
            return False
        listing.score = similarity.score(
            task.source_title, task.source_price,
            listing.title, listing.price, task.price_tolerance,
        )
        return listing.score >= task.similarity

File: src/avito/engine.cpp
#include "avito/engine.h"

#include <algorithm>

#include "avito/similarity.h"
#include "core/config.h"
#include "core/store.h"
#include "notify/notify.h"
#include "util/log.h"
#include "util/strings.h"

namespace avito {
namespace {

// Query widths tried when setting up a "similar listings" task, from precise to
// broad.
constexpr int kQueryWidths[] = {4, 3, 2};
constexpr size_t kEnoughNeighbours = 3;

std::string normalize_search_url(const std::string& url) {
    std::string text = util::trim(url);
    if (text.empty()) return {};
    if (!util::starts_with(text, "http")) {
        while (!text.empty() && text.front() == '/') text.erase(0, 1);
        text = "https://" + text;
    }

    size_t question = text.find('?');
    std::string base = question == std::string::npos ? text : text.substr(0, question);
    std::string query = question == std::string::npos ? "" : text.substr(question + 1);

    std::vector<std::string> kept;
    bool has_sort = false;
    for (const std::string& part : util::split(query, '&')) {
        if (part.empty()) continue;
        std::string name = part.substr(0, part.find('='));
        if (name == "p") continue;          // always the first page
        if (name == "s") has_sort = true;
        kept.push_back(part);
    }
    if (!has_sort) kept.push_back("s=104");  // newest first

    std::string result = base;
    if (!kept.empty()) result += "?" + util::join(kept, "&");
    return result;
}

}  // namespace

Engine::Engine(EngineListener* listener) : listener_(listener) {
    fetcher_.apply_settings(core::settings().get());
}

Engine::~Engine() {
    stop();
}

void Engine::start() {
    if (running_.load()) return;
    stop_.store(false);
    fetcher_.apply_settings(core::settings().get());
    running_.store(true);
    thread_ = std::thread(&Engine::loop, this);
    if (listener_) {
        listener_->on_running_changed(true);
        listener_->on_message("Слежение запущено");
    }
}

void Engine::stop() {
    if (!running_.load()) {
        if (thread_.joinable()) thread_.join();
        return;
    }
    stop_.store(true);
    wake_.notify_all();
    if (thread_.joinable()) thread_.join();
    running_.store(false);
    if (listener_) {
        listener_->on_running_changed(false);
        listener_->on_message("Слежение остановлено");
    }
}

void Engine::check_now(int task_id) {
    {
        std::lock_guard<std::mutex> lock(mutex_);
        manual_.push_back(task_id);
    }
    wake_.notify_all();
}

void Engine::schedule_soon(int task_id, double delay_seconds) {
    {
        std::lock_guard<std::mutex> lock(mutex_);
        next_run_[task_id] = core::now_seconds() + delay_seconds;
    }
    wake_.notify_all();
}

void Engine::forget_task(int task_id) {
    std::lock_guard<std::mutex> lock(mutex_);
    next_run_.erase(task_id);
}

void Engine::settings_changed() {
    fetcher_.apply_settings(core::settings().get());
    wake_.notify_all();
}

void Engine::loop() {
    while (!stop_.load()) {
        try {
            tick();
        } catch (const std::exception& error) {
            util::log_error("engine", std::string("Сбой в цикле слежения: ") + error.what());
        } catch (...) {
            util::log_error("engine", "Неизвестный сбой в цикле слежения");
        }

        std::unique_lock<std::mutex> lock(mutex_);
        wake_.wait_for(lock, std::chrono::seconds(2), [this] {
            return stop_.load() || !manual_.empty();
        });
    }
}

void Engine::tick() {
    // Manual checks jump the queue.
    while (true) {
        int task_id = 0;
        {
            std::lock_guard<std::mutex> lock(mutex_);
            if (manual_.empty()) break;
            task_id = manual_.front();
            manual_.pop_front();
        }
        core::Task task;
        if (core::store().task(task_id, task)) run_task(task);
        if (stop_.load()) return;
    }

    const double now = core::now_seconds();
    for (const core::Task& task : core::store().tasks()) {
        if (stop_.load()) return;
        if (!task.enabled) continue;

        double due;
        {
            std::lock_guard<std::mutex> lock(mutex_);
            auto it = next_run_.find(task.id);
            due = it != next_run_.end()
                      ? it->second
                      : task.last_check + std::max(kMinInterval, task.interval);
        }
        if (now >= due) run_task(task);
    }

    if (now - last_prune_ > 3600.0) {
        last_prune_ = now;
        core::Settings settings = core::settings().get();
        core::store().prune(settings.keep_history_days,
                            static_cast<size_t>(settings.max_feed_items) * 4);
    }
}

void Engine::run_task(core::Task task) {
    core::store().set_task_status(task.id, core::TaskStatus::Running, {});
    if (listener_) listener_->on_task_started(task.id);

    std::string error;
    core::TaskStatus status = core::TaskStatus::Ok;
    check_task(task, error, status);

    task.last_check = core::now_seconds();
    task.status = status;
    task.last_error = error;
    core::store().update_task(task);

    {
        std::lock_guard<std::mutex> lock(mutex_);
        next_run_[task.id] = task.last_check + std::max(kMinInterval, task.interval);
    }

    if (listener_) {
        listener_->on_task_finished(task.id, status, error);
        if (!error.empty()) listener_->on_message("«" + task.name + "»: " + error);
    }
}

bool Engine::check_task(core::Task& task, std::string& error, core::TaskStatus& status) {
    if (task.kind == core::TaskKind::Similar && task.derived_url.empty()) {
        if (!ensure_similar_ready(task, error)) {
            status = core::TaskStatus::Error;
            return false;
        }
    }

    const std::string url = normalize_search_url(task.search_url());
    if (url.empty()) {
        error = "У задачи не задана ссылка для поиска";
        status = core::TaskStatus::Error;
        return false;
    }

    net::FetchResult page = fetcher_.get(url, kSearchMarker, &stop_);
    if (!page.ok()) {
        error = page.error;
        status = page.status == net::FetchStatus::Blocked ? core::TaskStatus::Blocked
                                                          : core::TaskStatus::Error;
        return false;
    }
    util::log_info("engine", "Задача " + task.name + ": страница получена (" +
                                 page.transport + ")");

    std::vector<core::Listing> found = parse_search(page.html);
    if (found.empty()) {
        if (net::is_empty_result(page.html)) {
            // Nothing matches the query yet; that is a normal state.
            task.baseline_done = true;
            status = core::TaskStatus::Ok;
            return true;
        }
        error = "Авито вернул страницу без объявлений. Проверьте ссылку или попробуйте позже.";
        status = core::TaskStatus::Error;
        return false;
    }

    std::vector<core::Listing> candidates;
    std::vector<std::string> all_ids;
    all_ids.reserve(found.size());
    for (core::Listing& listing : found) {
        all_ids.push_back(listing.item_id);
        core::Listing copy = listing;
        if (passes_filters(task, copy)) candidates.push_back(std::move(copy));
    }

    // The first pass memorises what is already on Avito. Without it the user
    // would be flooded with notices about listings posted long ago.
    if (!task.baseline_done && !task.notify_existing) {
        core::store().mark_seen(task.id, all_ids);
        task.baseline_done = true;
        status = core::TaskStatus::Ok;
        if (listener_) {
            listener_->on_message("«" + task.name + "»: запомнил " +
                                  std::to_string(candidates.size()) +
                                  " текущих объявлений, жду новые");
        }
        return true;
    }
    task.baseline_done = true;

    std::vector<std::string> candidate_ids;
    candidate_ids.reserve(candidates.size());
    for (const core::Listing& listing : candidates) candidate_ids.push_back(listing.item_id);

    std::vector<std::string> new_ids = core::store().filter_new(task.id, candidate_ids);
    core::store().mark_seen(task.id, all_ids);

    if (new_ids.empty()) {
        status = core::TaskStatus::Ok;
        return true;
    }

    std::vector<core::Listing> fresh;
    const double stamp = core::now_seconds();
    for (core::Listing& listing : candidates) {
        if (std::find(new_ids.begin(), new_ids.end(), listing.item_id) == new_ids.end()) continue;
        listing.task_id = task.id;
        listing.task_name = task.name;
        listing.first_seen = stamp;
        fresh.push_back(listing);
    }
    if (fresh.empty()) {
        status = core::TaskStatus::Ok;
        return true;
    }

    task.found_total += static_cast<int>(fresh.size());
    core::store().add_listings(fresh);

    if (listener_) {
        listener_->on_new_listings(task.id, fresh);
        listener_->on_message("«" + task.name + "»: новых объявлений — " +
                              std::to_string(fresh.size()));
    }
    util::log_info("engine", "Задача " + task.name + ": новых объявлений " +
                                 std::to_string(fresh.size()));

    std::vector<std::string> errors =
        notify::dispatch(core::settings().get(), fresh, task.name);
    if (!errors.empty() && listener_) listener_->on_notify_errors(errors);

    status = core::TaskStatus::Ok;
    return true;
}

bool Engine::load_item_info(const std::string& url, ItemInfo& out, std::string& error) {
    net::FetchResult page = fetcher_.get(util::trim(url), {}, &stop_);
    if (!page.ok()) {
        error = page.error;
        return false;
    }
    if (is_removed_item(page.html)) {
        error =
            "Это объявление снято с публикации, взять его за образец нельзя. "
            "Выберите другое, похожее объявление.";
        return false;
    }
    out = parse_item(page.html, url);
    if (out.title.empty()) {
        error =
            "Не удалось прочитать объявление. Проверьте ссылку — нужна ссылка "
            "на страницу конкретного объявления.";
        return false;
    }
    return true;
}

bool Engine::ensure_similar_ready(core::Task& task, std::string& error) {
    ItemInfo info;
    if (!load_item_info(task.url, info, error)) return false;

    task.source_title = info.title;
    task.source_price = info.price;
    task.source_id = info.item_id;
    task.derived_url = pick_search_url(task, info);

    core::store().update_task(task);
    if (listener_) {
        listener_->on_message("«" + task.name + "»: образец — " + info.title +
                              " (запрос «" + task.derived_query + "»)");
    }
    return true;
}

std::string Engine::pick_search_url(core::Task& task, const ItemInfo& info) {
    // A query made of the whole title usually finds exactly one listing - the
    // sample. So, once, when the task is created, try shorter queries until
    // neighbours show up.
    std::string fallback_url;
    std::string fallback_query;

    for (int width : kQueryWidths) {
        std::string query = build_query(info.title, width);
        if (query.empty()) continue;

        std::string url = build_search_url(info, task.price_tolerance, query);
        if (fallback_url.empty()) {
            fallback_url = url;
            fallback_query = query;
        }

        net::FetchResult page = fetcher_.get(url, kSearchMarker, &stop_);
        if (!page.ok()) {
            util::log_info("engine", "Проба запроса «" + query + "» не удалась: " + page.error);
            continue;
        }

        size_t neighbours = 0;
        for (const core::Listing& listing : parse_search(page.html)) {
            if (listing.item_id != info.item_id) ++neighbours;
        }
        util::log_info("engine", "Запрос «" + query + "»: соседей " +
                                     std::to_string(neighbours));

        if (neighbours >= kEnoughNeighbours) {
            task.derived_query = query;
            return url;
        }
        fallback_url = url;
        fallback_query = query;
        if (stop_.load()) break;
    }

    task.derived_query = fallback_query;
    return fallback_url;
}

bool Engine::passes_filters(const core::Task& task, core::Listing& listing) const {
    const std::string title = util::to_lower(listing.title);
    for (const std::string& word : task.exclude_words()) {
        if (title.find(word) != std::string::npos) return false;
    }

    if (task.kind == core::TaskKind::Search) {
        if (task.price_min > 0 && listing.price > 0 && listing.price < task.price_min) return false;
        if (task.price_max > 0 && listing.price > 0 && listing.price > task.price_max) return false;
        return true;
    }

    // "Similar": never report the sample listing back to the user.
    if (!listing.item_id.empty() && listing.item_id == task.source_id) return false;

    listing.score = score(task.source_title, task.source_price, listing.title,
                          listing.price, task.price_tolerance);
    return listing.score >= task.similarity;
}

}  // namespace avito
